### PR TITLE
feat(forums): minimal forum-topics support (send_message topic_id + get_chat_info topics + forum message fields)

### LINKED ISSUE #10: Feature Request: First-class Support for Telegram Forum Topics (Threads)
### Is your feature request related to a problem?

Currently, `fast-mcp-telegram` treats messages primarily as a flat list. While `reply_to_msg_id` is supported, it is insufficient for robust workflows in **Telegram Forum Channels/Supergroups**.

Integrating AI agents into large communities requires the ability to:

1. **Identify context**: Know exactly which topic a message belongs to without manual "reply-chain" traversal.
2. **Target topics**: Send messages directly to a specific thread (topic) without needing an existing message ID to reply to.
3. **Filter history**: Read or search messages restricted to a specific `topic_id`.

In the current implementation, `build_message_result` only extracts the immediate `reply_to_msg_id`, and `search_messages_in_chat` lacks thread-level filtering.

#### 1. Enhanced Message Schema (Backward-Compatible)

Update `build_message_result` in `src/utils/message_format.py` to include:

* `topic_id`: The ID of the root message of the thread (often referred to as `reply_to_top_id` in Telethon/MTProto).
* `is_topic_message`: A boolean flag indicating the message belongs to a forum topic.
* `chat.is_forum`: A flag in the chat entity indicating if the chat has forum features enabled.

#### 2. Topic-Aware Messaging API

Extend `send_message` and `send_message_to_phone` in `src/tools/messages.py` with:

* An optional `topic_id` (or `message_thread_id`) parameter.
* **Logic**: If `topic_id` is provided, the server should internally construct an `InputReplyToMessage` with `top_msg_id` set to the `topic_id`. This allows sending messages to a topic even if no specific message is being replied to.

#### 3. Topic Filtering in Search and Read

Extend `search_messages_in_chat` in `src/tools/search.py` with:

* An optional `topic_id` filter.
* **Logic**: Pass this ID to Telethon’s `iter_messages(reply_to=topic_id)` in `_search_chat_messages_generator`. This natively restricts results to the specified thread.

#### 4. New Tool: `list_forum_topics`

Add a new tool to discover available topics in a chat:

* **Purpose**: Call `channels.getForumTopics`.
* **Output**: A list of topics including `id` (root message ID), `title`, and `icon_color`. This is essential for AI agents to "browse" the forum before interacting.

### Describe alternatives you've considered

One could use `invoke_mtproto` to manually call `channels.getForumTopics` or `messages.sendMessage` with custom `InputReplyToMessage`. However, this requires the AI agent to understand raw MTProto structures and `_` type hints, which is error-prone compared to having first-class tool support.

### Implementation Details (Suggested)

* **Metadata**: In `src/utils/message_format.py`, check `message.reply_to` for the `reply_to_top_id` attribute.
* **Search**: Update the `params` dict in `search_messages_impl` to include `topic_id` for logging and execution.
* **Entity**: Ensure `build_entity_dict` in `src/utils/entity.py` can detect the `forum` attribute on `Channel` objects.

### Why this matters

As Telegram Forum groups become the standard for large-scale community management, providing "topic-aware" context is critical for:

* **Thread Safety**: Preventing the AI from accidentally posting to the "General" branch.
* **Context Management**: Allowing agents to maintain separate conversation states for different topics within the same group.

### PR DESCRIPTION
**Pull request by Spiceman161**
Добавлена минимальная поддержка тем форумов в инструментах обмена сообщениями и получения информации о чатах.

Новые возможности:
- Разрешена отправка сообщений в конкретные темы форума с помощью необязательного параметра `topic_id` в `send_message`.
- Возврат компактных списков тем форума из `get_chat_info` для чатов с включённым форумом, с настраиваемым параметром `topics_limit`.

Улучшения:
- Поле `is_forum` стало доступно в сущностях чатов, а также включена метаинформация `topic_id` в результаты `message` и `edit_message` для сообщений в ветках форума; для нефорумных сообщений это поле опускается.

Документация:
- Задокументированы новые поля сообщений, связанные с форумом, и расширенные параметры `send_message` и `get_chat_info` в справочнике по инструментам.

Тесты:
- Добавлены целевые тесты, покрывающие извлечение метаданных тем, поведение списка тем форума в `get_chat_info`, приоритет `topic` по сравнению с `reply_to` при отправке, а также поведение `topic_id` при редактировании сообщений.

Fixes #10

## Summary by Sourcery

Добавлена минимальная поддержка тем форума во всех инструментах обмена сообщениями и информации о чатах, включая метаданные тем и определение форумов.

Новые возможности:
- Экспорт чатов с поддержкой форума через флаг `is_forum` в сущностях чатов.
- Возврат компактных списков тем форума с метаданными пагинации из `get_chat_info` для чатов с поддержкой форума.
- Возможность отправки сообщений в темы форума с помощью унифицированного параметра `reply_to` как ID корневого сообщения темы в `send_message`.
- Включение метаданных `topic_id` в результаты `message` и `edit_message` для сообщений в ветках форума.

Улучшения:
- Проверка и ограничение значений `topics_limit` при получении тем форума для обеспечения безопасных значений по умолчанию и допустимых границ.
- Сделать ошибки при получении списка тем форума некритичными в `get_chat_info`, чтобы основная информация о чате оставалась доступной.
- Приведение моков клиентского соединения и внутренних вспомогательных функций в соответствие с ожиданиями API Telethon.

Документация:
- Документировать связанные с форумами поля сообщений, `is_forum`, `topics`/`topics_has_more`, а также обновлённые параметры `send_message` и `get_chat_info` в справочнике по инструментам.
- Обновить основной список возможностей в README, чтобы упомянуть обмен сообщениями с учётом форумов, определение `is_forum` и возможности получения списка тем.
- Расширить документацию по тестам в CONTRIBUTING командами для специализированного набора тестов по темам форума и необязательным живым интеграционным тестам.

Тесты:
- Добавить отдельный модуль тестов для тем форума, охватывающий извлечение метаданных тем, поведение `send`/`edit` в ветках форума, семантику пагинации списка тем и обработку ошибок.
- Обновить общие фикстуры тестов и добавить необязательные живые интеграционные тесты для проверки реальных API тем форума.

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Add minimal forum topics support across messaging tools and chat info, including topic metadata and forum detection.

New Features:
- Expose forum-enabled chats via an is_forum flag on chat entities.
- Return compact forum topic lists with pagination metadata from get_chat_info for forum-enabled chats.
- Allow posting into forum topics by using the unified reply_to parameter as a topic root message ID in send_message.
- Include topic_id metadata in message and edit_message results for messages in forum threads.

Enhancements:
- Validate and clamp topics_limit values when fetching forum topics to ensure safe defaults and bounds.
- Make forum topic listing failures non-fatal in get_chat_info to keep core chat info available.
- Align client connectivity mocks and internal helpers with Telethon API expectations.

Documentation:
- Document forum-related message fields, is_forum, topics/topics_has_more, and the updated send_message and get_chat_info parameters in the tools reference.
- Update the main README feature list to mention forum-aware messaging, is_forum detection, and topic listing capabilities.
- Extend CONTRIBUTING test documentation with commands for the dedicated forum topics test suite and optional live integration tests.

Tests:
- Add a dedicated forum topics test module covering topic metadata extraction, send/edit behavior in forum threads, topic list pagination semantics, and error handling.
- Update shared test fixtures and add optional live integration tests to validate real-world forum topic APIs.

</details>

Новые возможности:
- Разрешить отправку сообщений в конкретные форумные топики через необязательный параметр `topic_id` в `send_message`.
- Предоставлять компактные списки форумных топиков и метаданные пагинации из `get_chat_info` для чатов с включёнными форумами.
- Включать метаданные `topic_id` в результаты `message` и `edit_message` для сообщений в ветках форума.
- Экспонировать флаг `is_forum` у сущностей чатов для идентификации супергрупп с включёнными форумами.

Улучшения:
- Отдавать приоритет ответу на конкретное сообщение по сравнению с привязкой к форумному топику, когда одновременно переданы `reply_to_msg_id` и `topic_id`.
- Ограничивать и валидировать `topics_limit` при получении форумных топиков, чтобы обеспечить разумные значения по умолчанию и границы.

Документация:
- Задокументировать связанные с форумами поля сообщений, новый параметр `topic_id` для `send_message`, а также `topics_limit` и форумные поля в `get_chat_info` в разделе справки по инструментам.
- Добавить README по тестированию с инструкциями по запуску нового набора тестов для форумных топиков и необязательных живых интеграционных тестов.

Тесты:
- Добавить расширенный набор тестов для форумных топиков, охватывающий извлечение метаданных топиков, приоритет между отправкой и ответом, поведение форумных топиков в `get_chat_info`, крайние случаи в списках топиков и обработку `topic_id` при редактировании сообщений.
- Адаптировать тестовые фикстуры к новому поведению клиента и добавить необязательное живое интеграционное покрытие для форумных топиков.

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Добавлена минимальная поддержка тем форума во всех инструментах обмена сообщениями и информации о чатах, включая метаданные тем и определение форумов.

Новые возможности:
- Экспорт чатов с поддержкой форума через флаг `is_forum` в сущностях чатов.
- Возврат компактных списков тем форума с метаданными пагинации из `get_chat_info` для чатов с поддержкой форума.
- Возможность отправки сообщений в темы форума с помощью унифицированного параметра `reply_to` как ID корневого сообщения темы в `send_message`.
- Включение метаданных `topic_id` в результаты `message` и `edit_message` для сообщений в ветках форума.

Улучшения:
- Проверка и ограничение значений `topics_limit` при получении тем форума для обеспечения безопасных значений по умолчанию и допустимых границ.
- Сделать ошибки при получении списка тем форума некритичными в `get_chat_info`, чтобы основная информация о чате оставалась доступной.
- Приведение моков клиентского соединения и внутренних вспомогательных функций в соответствие с ожиданиями API Telethon.

Документация:
- Документировать связанные с форумами поля сообщений, `is_forum`, `topics`/`topics_has_more`, а также обновлённые параметры `send_message` и `get_chat_info` в справочнике по инструментам.
- Обновить основной список возможностей в README, чтобы упомянуть обмен сообщениями с учётом форумов, определение `is_forum` и возможности получения списка тем.
- Расширить документацию по тестам в CONTRIBUTING командами для специализированного набора тестов по темам форума и необязательным живым интеграционным тестам.

Тесты:
- Добавить отдельный модуль тестов для тем форума, охватывающий извлечение метаданных тем, поведение `send`/`edit` в ветках форума, семантику пагинации списка тем и обработку ошибок.
- Обновить общие фикстуры тестов и добавить необязательные живые интеграционные тесты для проверки реальных API тем форума.

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Add minimal forum topics support across messaging tools and chat info, including topic metadata and forum detection.

New Features:
- Expose forum-enabled chats via an is_forum flag on chat entities.
- Return compact forum topic lists with pagination metadata from get_chat_info for forum-enabled chats.
- Allow posting into forum topics by using the unified reply_to parameter as a topic root message ID in send_message.
- Include topic_id metadata in message and edit_message results for messages in forum threads.

Enhancements:
- Validate and clamp topics_limit values when fetching forum topics to ensure safe defaults and bounds.
- Make forum topic listing failures non-fatal in get_chat_info to keep core chat info available.
- Align client connectivity mocks and internal helpers with Telethon API expectations.

Documentation:
- Document forum-related message fields, is_forum, topics/topics_has_more, and the updated send_message and get_chat_info parameters in the tools reference.
- Update the main README feature list to mention forum-aware messaging, is_forum detection, and topic listing capabilities.
- Extend CONTRIBUTING test documentation with commands for the dedicated forum topics test suite and optional live integration tests.

Tests:
- Add a dedicated forum topics test module covering topic metadata extraction, send/edit behavior in forum threads, topic list pagination semantics, and error handling.
- Update shared test fixtures and add optional live integration tests to validate real-world forum topic APIs.

</details>

</details>

### CODE REVIEW

File: tests/test_forum_topics_minimal.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
+from datetime import UTC, datetime
+from types import SimpleNamespace
+from unittest.mock import AsyncMock, patch
+
+import pytest
+
+from src.tools.contacts import get_chat_info_impl
+from src.tools.messages import _send_message_or_files
+from src.utils.message_format import build_message_result
```
[the remark this answers is not in the document]
Comment (leshchenko1979) [reply tests/test_forum_topics_minimal.py]: Besides what Sourcery had found, can you also fix edit_message? It should also be topic_aware.

Also, can you explain the difference between topic_id and top_msg_id? I believe they are the same, aren't they?

[the remark this answers is not in the document]
Comment (Spiceman161) [reply tests/test_forum_topics_minimal.py]: С Вашего позволения перейду на русский)

Спасибо за ревью!

Замечания Sourcery исправлю.

По вашим вопросам:

1) edit_message и topic-aware

Согласен, нужно довести до консистентности с forum-support.

Предлагаю минимальный, backward-compatible вариант:

1. Оставить текущий интерфейс рабочим как есть.
2. Добавить в результат edit_message те же forum-поля (topic_id, top_msg_id) для форумных сообщений.
3. (Опционально) добавить topic_id в edit_message как guard-параметр: если передан, проверяем, что редактируемое сообщение принадлежит этому topic/thread; иначе возвращаем ошибку.

Технически для самого редактирования message_id в чате уже достаточно, но guard + метаданные сделают поведение действительно topic-aware на уровне API-контракта.

2) Разница между topic_id и top_msg_id

Да, в большинстве forum-кейсов они совпадают.

• top_msg_id — нативное поле Telegram (id корневого сообщения треда).
• topic_id — нормализованное API-поле для удобства клиентов (thread/topic id).

То есть обычно одинаковое значение, но top_msg_id — “сырой источник”, а topic_id — клиентский унифицированный контракт.

Пусть будут оба или оставить только top_msg_id?

**Comment by leshchenko1979** [comment]
Я прошу прощения - сейчас у нас один PR на много задач. Получается расфокусировка. Я предлагаю сконцентрироваться тут на forum topics.

Важной целью для меня остаётся оставить интерфейсы инструментов простыми, чтобы не увеличивать контекстное окно.

Поэтому:

1. Давайте, плз, уберем все, что не связано с forum topics.

2. Я был неправ по поводу использования параметра topic_id в edit_message - его там быть не должно, номера сообщения достаточно для его идентификации для этого инструмента.

3. При возврате message entity давать только topic_id, без top_msg_id.

4. При перечислении forum topics в результатах get_chat_id давать только topic_id, без top_msg_id.

**Comment by Spiceman161** [comment]
Я обновил PR по вашим пунктам:

1) Убрал все не относящееся к forum topics.
2) Убрал параметр topic_id из edit_message (оставили простой интерфейс: chat_id + message_id достаточно).
3) В message entity теперь возвращается только topic_id (без top_msg_id).
4) В get_chat_info для forum topics теперь компактный формат topics: [{topic_id, title}].

Также обновил тесты и документацию под эту модель.
Локально: pytest (full) + ruff проходят.

Если все ок, буду рад финальному ревью.

File: src/utils/message_format.py [hunk 2]
```diff
@@ -399,12 +399,24 @@ async def build_message_result(
         "sender": sender,
     }
 
+    reply_to = getattr(message, "reply_to", None)
```
Comment (leshchenko1979) [inline src/utils/message_format.py]: А это изменение зачем?

File: docs/Tools-Reference.md
Comment (leshchenko1979) [inline docs/Tools-Reference.md:1]: Please review all of your changes and update this tool reference accordingly. We have changed 2 tools and 2 entities - this should be reflected here.

**Review by leshchenko1979 — CHANGES_REQUESTED** [review]
Thanks, we are almost there - requested a couple of minor changes.

**Comment by leshchenko1979** [comment]
One more thing.

When retreiving search results, we return the field `has_more` which helps the agent understand if there are more results beyond the given limit.

Can we do the same for forum topics and `get_chat_info`?

**Comment by leshchenko1979** [comment]
@sourcery-ai review

File: src/tools/contacts.py [hunk 3]
```diff
@@ -207,18 +208,52 @@ async def find_chats_impl(
...
+    raw_topics = getattr(result, "topics", []) or []
+    topics = []
+    for topic in raw_topics:
+        topic_id = getattr(topic, "id", None)
+        title = getattr(topic, "title", None)
+        if topic_id is None or title is None:
+            continue
+        topics.append({"topic_id": topic_id, "title": title})
+
+    return {"topics": topics, "has_more": len(raw_topics) >= safe_limit}
```
Comment (leshchenko1979) [inline src/tools/contacts.py]: Если в чате ровно 20 тем, и `safe_limit` = 20, то такая реализация ошибочно даст `has_more` = True. Нужно добавить тесты на edge cases.

File: src/tools/contacts.py [hunk 4]
```diff
@@ -230,8 +265,18 @@ async def get_chat_info_impl(chat_id: str) -> dict[str, Any]:
...
-    # Return enriched info with counts and about/bio when applicable
-    return await build_entity_dict_enriched(entity)
+    info = await build_entity_dict_enriched(entity)
+
+    # Add topics list only for forum-enabled chats.
+    if info and info.get("is_forum"):
+        try:
+            topics_result = await _list_forum_topics(entity, topics_limit)
+            info["topics"] = topics_result["topics"]
+            info["topics_has_more"] = topics_result["has_more"]
```
Comment (leshchenko1979) [inline src/tools/contacts.py:301]: Ненужное дублирование

File: src/tools/messages.py [hunk 5]
```diff
@@ -385,8 +390,10 @@ def _extract_send_message_params(
         "message": message,
         "message_length": len(message),
         "reply_to_msg_id": reply_to_msg_id,
+        "topic_id": topic_id,
         "parse_mode": parse_mode,
         "has_reply": reply_to_msg_id is not None,
+        "has_topic": topic_id is not None,
```
Comment (leshchenko1979) [inline src/tools/messages.py]: Не нужно - если topic_id присутствует, то сообщение внутри темы, иначе - вне темы.

File: src/tools/messages.py [hunk 6]
```diff
@@ -449,10 +458,11 @@ async def send_message_impl(
         message,
         files,
         reply_to_msg_id,
+        topic_id,
         resolved_parse_mode,
         "send_message",
         _extract_send_message_params(
-            chat_id, message, reply_to_msg_id, parse_mode, files
+            chat_id, message, reply_to_msg_id, topic_id, parse_mode, files
```
Comment (leshchenko1979) [inline src/tools/messages.py]: Перейти на поименованные параметры. С таким количество параметров это будет безопаснее.

File: tests/test_forum_topics_minimal.py
Comment (leshchenko1979) [inline tests/test_forum_topics_minimal.py:1]: Добавить больше edge cases

**Comment by leshchenko1979** [comment]
@sourcery-ai review

File: src/tools/contacts.py [hunk 7]
```diff
@@ -207,18 +208,52 @@ async def find_chats_impl(
...
+
+    raw_topics = getattr(result, "topics", []) or []
+    topics = []
+    for topic in raw_topics:
+        topic_id = getattr(topic, "id", None)
+        title = getattr(topic, "title", None)
+        if topic_id is None or title is None:
+            continue
+        topics.append({"topic_id": topic_id, "title": title})
+
```
[the remark this answers is not in the document]
Comment (leshchenko1979) [reply src/tools/contacts.py:265]: Самый лучший способ - запрашивать у Телеграма на 1 больше топиков, чем лимит, запрошенный пользователем, и смотреть, удалось ли получить больше топиков или нет. 

Но, опять же, в каких то случаях Телеграм может отдать страницу меньшим размером, чем запрошенный, и такие случаи тоже должна быть отработаны.

И хорошо бы было включить интегральный тест на живых данных, чтобы посмотреть, что на самом деле возвращает API Телеграма, и работает ли наш код вживую.

**Comment by Spiceman161** [comment]
Вроде закрыл все оставшиеся замечания:

1. topic_id vs top_msg_id

Убрал top_msg_id из API-ответов, оставил только topic_id как единый нормализованный идентификатор темы. reply_to_top_id используется внутри как источник, но наружу не торчит.

2. edit_message — topic-aware

Реализовано: вынес _extract_topic_metadata(message) в общий хелпер в message_format.py, и edit_message_impl теперь тоже вызывает его через result.update(...). Логика дедуплицирована, поведение send/edit консистентно. Тесты покрывают все случаи: с reply_to_top_id, fallback через reply_to_msg_id, нет топика, не форум.

3. Изменение в message_format.py:416

Это был косметический рефакторинг — промежуточная переменная reply_to, чтобы избежать вложенного getattr(getattr(...)). Никакой функциональной нагрузки не несла. Откатил обратно к оригинальному виду, чтобы не увеличивать diff без причины.

4. docs/Tools-Reference.md

Обновлён в коммите c5341c8:

• send_message: добавлен параметр topic_id с примером отправки в тему
• get_chat_info: добавлен параметр topics_limit, задокументированы возвращаемые поля (is_forum, topics, topics_has_more)
• Uniform Entity Schema: добавлен флаг is_forum

**Comment by leshchenko1979** [comment]
We still have 6 unresolved comments. Do you see them in the`Files changed` section? @Spiceman161

File: src/server_components/tools_register.py [hunk 8]
```diff
@@ -152,6 +152,7 @@ async def send_message(
         chat_id: str,
         message: str,
         reply_to_msg_id: int | None = None,
+        topic_id: int | None = None,
```
Comment (leshchenko1979) [inline src/server_components/tools_register.py]: I've been thinking about this adding another parameter.

In practice, you can't answer a message and a topic at the same time.

So, if both these args are given, they may conflict each other.

On the other hand, if you just change the existing `reply_to_msg_id` to `reply_to` and write in the doc string that either a message or a topic id may be given here, wouldn't it be cleaner?

What do you think?

**Review by leshchenko1979 — CHANGES_REQUESTED** [review]

**Comment by Spiceman161** [comment]
1) has_more / пагинация forum topics

Сделал переработку логики has_more в _list_forum_topics:

• запрашиваем у Telegram limit + 1 (c clamp до 100),
• наружу возвращаем только первые limit,
• has_more считаем по факту получения “лишнего” элемента (и учитываем метаданные ответа, где применимо),
• добавил защиту от невалидного topics_limit.

Это закрывает кейс с ложным has_more=true, когда тем ровно limit.

───

2) Чистка/упрощение API и кода

По замечаниям про избыточность:

• убрал мелкое дублирование в get_chat_info_impl (ошибка not-found формируется один раз),
• убрал лишний служебный флаг has_topic из params (избыточен, т.к. достаточно topic_id).

───

3) Тесты: edge-cases + интеграционный прогон

Расширил tests/test_forum_topics_minimal.py:

• edge-cases для _list_forum_topics (limit, limit+1, clamp, invalid limit, фильтрация битых тем),
• дополнительные сценарии для get_chat_info_impl (not found / non-fatal failure topic listing),
• дополнительные сценарии для topic metadata extraction.

Также добавлен opt-in integration test (@pytest.mark.integration) на живом API:

• тест выключен по умолчанию (чтобы CI не был flaky и не зависел от внешней среды),
• запускается только при явных env-переменных.

Обезличенный результат live-интеграционного прогона

Live sample (anonymized):
```
topics_count_returned: 4
has_more: false
topics: [{"topic_id": 117, "title": "Topic A"}, {"topic_id": 52, "title": "Topic B"}, {"topic_id": 1, "title": "General"}, {"topic_id": 3, "title": "Topic C"}]
```

Локально полный прогон после изменений:

`197 passed, 1 skipped`

1 skipped — это ожидаемо: это opt-in integration path (или другой environment-dependent skip), который в дефолтном CI-режиме не должен падать при отсутствии внешней конфигурации.

File: src/server_components/tools_register.py (hunk 8 above)
Comment (Spiceman161) [reply src/server_components/tools_register.py]: I have not changed this part yet in code, because we should align on your preferred contract first.

My current view: a single reply_to can be cleaner, but only if we make the semantics explicit for LLM usage.

To reduce ambiguity, I suggest these safeguards:

• Strict schema: prefer structured input like
reply_to: { kind: "message" | "topic", id: number }
instead of a plain integer.
• Server-side validator: reject ambiguous/invalid combinations with deterministic, actionable errors.
• LLM-focused docs/examples: include clear positive + negative examples (message reply vs topic targeting).
• Tool-description constraints: explicitly state what must not be combined and what fallback behavior is expected.

With this setup, confusion risk is low. In practice, residual mistakes are mostly expected from small/cheap or older models — and those are unlikely to be the primary users of a more advanced tool like fast-mcp-telegram anyway 🙂

If you agree, I can prepare a focused follow-up patch for the unified reply_to API with these guardrails included.

File: src/tools/contacts.py [hunk 9]
```diff
@@ -207,31 +208,87 @@ async def find_chats_impl(
...
+    topics = []
+    for topic in raw_topics[:requested_limit]:
+        topic_id = getattr(topic, "id", None)
+        title = getattr(topic, "title", None)
+        if topic_id is None or title is None:
+            continue
+        topics.append({"topic_id": topic_id, "title": title})
+
+    return {"topics": topics, "has_more": has_more}
```
Comment (leshchenko1979) [inline src/tools/contacts.py:266]: Если limit=100 и в чате больше 100 тем, эта реализация ошибочно даст has_more=False.

File: tests/README.md
Comment (leshchenko1979) [inline tests/README.md:1]: Это ai slop, лишняя информация для Readme. Убрать.

File: tests/test_forum_topics_minimal.py
Comment (leshchenko1979) [reply tests/test_forum_topics_minimal.py:1]: Добавить кейс, если в форуме больше 100 тем, а запрошено 100.

File: src/server_components/tools_register.py (hunk 8 above)
Comment (leshchenko1979) [reply src/server_components/tools_register.py]: Я бы не хотел делать `reply_to` сложным словарем - так ещё больше вероятность, что агенты будут ошибаться.

Предлагаю оставить параметр с коротким названием `reply_to: int | None` и дать пояснения в docstring.

**Review by leshchenko1979 — CHANGES_REQUESTED** [review]

**Comment by leshchenko1979** [comment]
@sourcery-ai review

**Review by SourceryAI — COMMENTED** [review]
Привет — я оставил несколько общих замечаний:

- В `_extract_send_message_params` значение `has_reply` по‑прежнему определяется только на основе `reply_to_msg_id`, поэтому вызовы, которые используют только `topic_id`, будут устанавливать `reply_to` внутри, но в логах/метриках выглядеть как `has_reply=False`. Имеет смысл обновить этот флаг (или добавить отдельный), чтобы наблюдаемость соответствовала фактическому поведению.
- В тесте метод `TestClient.is_connected` был изменён с `async def` на синхронный метод. Если в продакшене код делает `await client.is_connected()`, тесты теперь будут использовать другой интерфейс и могут скрыть проблемы — либо сохраните сигнатуру неизменной, либо соответственно скорректируйте места вызова.
- В примерах Tools-Reference для `get_chat_info` есть два подряд идущих примера с идентичными параметрами: один с подписью «Get channel information», другой — «Get forum topics». Стоит либо дифференцировать первый (как чисто профильную информацию), либо второй (с явным `topics_limit`), чтобы в примерах наглядно показывалось новое поведение.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- In `_extract_send_message_params`, `has_reply` is still derived solely from `reply_to_msg_id`, so calls that use only `topic_id` will set `reply_to` under the hood but appear as `has_reply=False` in logs/metrics; consider updating this flag (or adding a separate one) so observability matches actual behavior.
- The test `TestClient.is_connected` method was changed from `async def` to a synchronous method; if production code awaits `client.is_connected()`, tests will now exercise a different interface and may mask issues—either keep the signature consistent or adjust calling sites accordingly.
- In the Tools-Reference examples for `get_chat_info`, there are two consecutive examples with identical params, one labeled "Get channel information" and one "Get forum topics"; consider differentiating the first (pure profile info) or the second (with an explicit `topics_limit`) so the examples clearly show the new behavior.
~~~

</details>

***

Привет, @leshchenko1979! 👋

Спасибо, что попробовали Sourcery, оставив комментарий `@sourcery-ai review`! 🚀

Установите [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new), чтобы получать автоматические ревью кода для каждого pull request ✨

<sub>
Помогите мне стать полезнее! Поставьте, пожалуйста, 👍 или 👎 к каждому комментарию, и я буду использовать этот фидбек, чтобы улучшать ревью.
</sub>

<details>
<summary>Original comment in English</summary>

Hey - I've left some high level feedback:

- In `_extract_send_message_params`, `has_reply` is still derived solely from `reply_to_msg_id`, so calls that use only `topic_id` will set `reply_to` under the hood but appear as `has_reply=False` in logs/metrics; consider updating this flag (or adding a separate one) so observability matches actual behavior.
- The test `TestClient.is_connected` method was changed from `async def` to a synchronous method; if production code awaits `client.is_connected()`, tests will now exercise a different interface and may mask issues—either keep the signature consistent or adjust calling sites accordingly.
- In the Tools-Reference examples for `get_chat_info`, there are two consecutive examples with identical params, one labeled "Get channel information" and one "Get forum topics"; consider differentiating the first (pure profile info) or the second (with an explicit `topics_limit`) so the examples clearly show the new behavior.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- In `_extract_send_message_params`, `has_reply` is still derived solely from `reply_to_msg_id`, so calls that use only `topic_id` will set `reply_to` under the hood but appear as `has_reply=False` in logs/metrics; consider updating this flag (or adding a separate one) so observability matches actual behavior.
- The test `TestClient.is_connected` method was changed from `async def` to a synchronous method; if production code awaits `client.is_connected()`, tests will now exercise a different interface and may mask issues—either keep the signature consistent or adjust calling sites accordingly.
- In the Tools-Reference examples for `get_chat_info`, there are two consecutive examples with identical params, one labeled "Get channel information" and one "Get forum topics"; consider differentiating the first (pure profile info) or the second (with an explicit `topics_limit`) so the examples clearly show the new behavior.
~~~

</details>

***

Hi @leshchenko1979! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

</details>

File: tests/README.md
Comment (Spiceman161) [reply tests/README.md:1]: README сокращён

File: src/server_components/tools_register.py (hunk 8 above)
Comment (Spiceman161) [reply src/server_components/tools_register.py]: По reply_to оставил упрощённый контракт в духе ваших требований (без сложного словаря), чтобы снизить шанс ошибок у агентов.
Рекомендации Sourcery учел частично: улучшил observability (has_reply учитывает и topic_id) и расширил тесты.
Более строгую reply-модель (если понадобится) думаю лучше вынести в отдельный follow-up PR, чтобы не раздувать текущий scope.

**Comment by Spiceman161** [comment]
Сделал opt-in live проверку на реальной форум-группе (ID скрыт).

CHAT_ID=<redacted>, initial_topics_count=3

[PHASE A] before mass-create

• ProbeLog(requested_limit=20, returned_topics_count=4, first_topic_id=117, last_topic_id=3, has_more=False, probe_executed=False)
• ProbeLog(requested_limit=100, returned_topics_count=4, first_topic_id=117, last_topic_id=3, has_more=False, probe_executed=False)

Довёл до >100:

• created_topics=98
• total_topics_now=101
• after_count=101

[PHASE B] after ensure >100

• ProbeLog(requested_limit=20, returned_topics_count=20, first_topic_id=223, last_topic_id=204, has_more=True, probe_executed=False)
• ProbeLog(requested_limit=100, returned_topics_count=100, first_topic_id=223, last_topic_id=52, has_more=True, probe_executed=True)

Это подтверждает корректную работу has_more для границ 20 и 100, включая ключевой кейс limit=100 при >100 тем.

File: src/tools/contacts.py (hunk 9 above)
Comment (Spiceman161) [reply src/tools/contacts.py:266]: Исправил этот кейс в ce919e4:

• для topics_limit < 100 используется overfetch (limit+1),
• для topics_limit = 100 (cap) добавлен probe-запрос следующей страницы (limit=1 с offset_topic=last_topic_id).

Добавлены edge-тесты и live-проверка на реальном форуме (>100 тем), где has_more=true и probe_executed=true.

File: tests/test_forum_topics_minimal.py
Comment (Spiceman161) [reply tests/test_forum_topics_minimal.py:1]: Добавил дополнительные edge-cases в 4b0dcf0, включая граничные сценарии для limit=100.
Плюс сделан opt-in live run на реальном форуме (анонимизированный лог уже приложен выше), где подтверждён кейс limit=100 при >100 тем.

File: tests/README.md
Comment (leshchenko1979) [reply tests/README.md:1]: Этой информации не место в README.md. Если хочется добавить информацию о тестах в документацию, для этого есть CONTRIBUTING.md.

File: src/server_components/tools_register.py (hunk 8 above)
Comment (leshchenko1979) [reply src/server_components/tools_register.py]: topic_id не нужен - вместо него есть reply_to

File: docs/Tools-Reference.md [hunk 10]
```diff
@@ -80,7 +80,8 @@ All tools return chat/user objects in the same schema via `build_entity_dict`:
   "first_name": "John",         // users
   "last_name": "Doe",           // users
   "members_count": 1234,          // groups (when available)
-  "subscribers_count": 56789      // channels (when available)
+  "subscribers_count": 56789,     // channels (when available)
+  "is_forum": true               // present and true for forum-enabled supergroups only
 }
```
Comment (leshchenko1979) [inline docs/Tools-Reference.md:85]: А как же topics и topics_has_more?

File: docs/Tools-Reference.md
Comment (leshchenko1979) [inline docs/Tools-Reference.md:1]: Нужно привести в соответствие после окончания исправления сигнатуры send_message

**Review by leshchenko1979 — CHANGES_REQUESTED** [review]

File: tests/README.md
Comment (leshchenko1979) [reply tests/README.md:1]: А вот что реально стоит обновить в README.md, так это разделы Features и Available Tools.

Comment (Spiceman161) [reply tests/README.md]: Удалил tests/README.md, а тестовую документацию перенёс в CONTRIBUTING.md (включая запуск forum/live integration тестов).

Дополнительно обновил корневой README.md в разделах Features/Available Tools, чтобы он отражал актуальный контракт (send_message.reply_to) и forum topics (topics, topics_has_more).

**Comment by leshchenko1979** [comment]
@SourceryAI look for excessive or stinky code and look for optimizations

**Review by leshchenko1979 — APPROVED** [review]

**Comment by leshchenko1979** [comment]
Merged, thanks for your work @Spiceman161 !!!
